Fix/refactor codebase

CihatAltiparmak opened the pull request.

CihatAltiparmak committed:
Added config files

CihatAltiparmak committed:
Beginning of Middleware Benchmark Implementation

CihatAltiparmak committed:
Refactor benchmark code

- Refactored test_scenario_perception_pipeline.cpp
- Added event handler because of interaction problem between panda_arm_spawner and test_scenario_perception_pipeline node

CihatAltiparmak committed:
Integrated ros-industrial-ci and updated some external dependencies

CihatAltiparmak committed:
Changed the function namings by camelCase and some changes in launch files

CihatAltiparmak commented on launch/test_scenario_perception_pipeline.launch.py:
```diff
@@ -8,10 +8,24 @@
 from ament_index_python.packages import get_package_share_directory
 from moveit_configs_utils import MoveItConfigsBuilder
 from launch_param_builder import ParameterBuilder
+from launch.actions import (
+    DeclareLaunchArgument,
+    EmitEvent,
+    ExecuteProcess,
+    LogInfo,
+    RegisterEventHandler,
+    TimerAction,
+)
+from launch.event_handlers import (
+    OnExecutionComplete,
+    OnProcessExit,
+    OnProcessIO,
+    OnProcessStart,
+    OnShutdown,
+)
```
Remove unnecessary imports

CihatAltiparmak commented on test/test_scenario_perception_pipeline.cpp:
```diff
@@ -10,89 +10,145 @@
+      failure_number++;
+  }
+
+  return { success_number, failure_number };
```
Add this metrics to benchmark

CihatAltiparmak commented on test/test_scenario_perception_pipeline.cpp:
```diff
@@ -10,89 +10,145 @@
+class ScenarioPerceptionPipelineTestCaseCreator
+{
+private:
+  static inline std::vector<std::vector<geometry_msgs::msg::Pose>> test_cases_ = {};
```
Is this part okay or do we need to get better implementation?

CihatAltiparmak replied:
I also need some insights about this part, @sjahr, @henningkayser

CihatAltiparmak commented on test/test_scenario_perception_pipeline.cpp:
```diff
@@ -10,89 +10,145 @@
+
+BENCHMARK_DEFINE_F(ScenarioPerceptionPipelineFixture, test_scenario_perception_pipeline)(benchmark::State& st)
+{
+  auto selected_test_case = ScenarioPerceptionPipelineTestCaseCreator::selectTestCases(0);
```
Test case selection must be based on given arguments to benchmark class.

CihatAltiparmak replied:
Dear @henningkayser  and @sjahr, I actually solved this issue by giving this argument in launch file. But i need some insights related to this part.

CihatAltiparmak commented on test/test_scenario_perception_pipeline.cpp:
```diff
@@ -10,89 +10,145 @@
+    target_pose2_msg.orientation.w = 1.0;
+    target_pose2_msg.position.x = 0.5;
+    target_pose2_msg.position.y = -0.5;
+    target_pose2_msg.position.z = 0.7;
```
These test cases must be fetched from parameter files.

CihatAltiparmak committed:
Made test case creation flexible and removed unnecessary imports from launch file

CihatAltiparmak committed:
Added how to run and how to install sections to README and some pre-fix

CihatAltiparmak commented:
Dear @henningkayser  and @sjahr ,

I need some review for beginning of implementations. I already reviewed some parts before. While reviewing, it would be great to review all parts of code, not changed parts at the moment. Thank you in advance for your efforts :pray:

sjahr commented on package.xml:
```diff
@@ -12,6 +12,9 @@
   <depend>benchmark</depend>
+  <depend>dynmsg</depend>
+  <depend>nav_msgs</depend>
+  <depend>ament_index_cpp</depend>
```
Would it make sense to add the rmw implementations here as dependencies?

CihatAltiparmak replied:
It absolutely makes sense. I just missed this part. I fixed this new commit. Thank you.

sjahr commented on test/test_scenario_perception_pipeline.cpp:
```diff
@@ -7,92 +7,165 @@
-        }
-    }
+  moveit::planning_interface::MoveGroupInterface::Plan plan_msg;
+  const auto ok = static_cast<bool>(move_group_interface_ptr_->plan(plan_msg));
```
That static_cast shouldn't be necessary because the result class has a bool operator. I think you could just put `move_group_interface_ptr_->plan(plan_msg)` into the if statement without temporary variables

CihatAltiparmak replied:
I have stolen [here](https://moveit.picknik.ai/main/doc/tutorials/your_first_project/your_first_project.html#plan-and-execute-using-movegroupinterface) :see_no_evil: . But it is bad practice. I have yet to solve this part but i will. We must fix this bad practice in documentation as well.

sjahr commented on test/test_scenario_perception_pipeline.cpp:
```diff
@@ -7,92 +7,165 @@
 
+  if (ok)
+  {
+    move_group_interface_ptr_->execute(plan_msg);
```
Do we need the execution at all? If not you could just return `move_group_interface_ptr_->plan(plan_msg)` with from this function

CihatAltiparmak replied:
I couldn't understand your question. Wasn't our goal to test all pipeline rather than just planning pipeline? By the way i think the effectively measuring of middleware benchmarking is proportional to data-transfering works. I'm wondering your ideas about all of it.

sjahr commented on test/test_scenario_perception_pipeline.cpp:
```diff
@@ -7,92 +7,165 @@
 #include <moveit/moveit_cpp/planning_component.h>
```
I think you should separate the declarations form the implementations by adding header files for your classes. That way everything would be more cleaned up and easier to work with. I'd propose this repository structure:

```
launch/
include/
 scenario_perception_pipeline.hpp --> Contains class definitions
src/
  scenario_perception_pipeline.cpp --> Contains all the class definitions
  benchmark_main.cpp --> contains just the main function
test/ --> Usually this contains the unittests (you can neglect it for now until we get to the point where we write unittests)
```

CihatAltiparmak replied:
I refactored based on your review with a couple of details. Could you take a look at it and then give a feedback again?

sjahr commented on test/test_scenario_perception_pipeline.cpp:
```diff
@@ -7,92 +7,165 @@
+const std::string PLANNING_GROUP = "panda_arm";
 
-class ScenarioPerceptionPipeline {
+class ScenarioPerceptionPipeline
```
In the header file you should add documentation for every class and (member) function you create to make it easier to understand

sjahr commented on test/test_scenario_perception_pipeline.cpp:
```diff
@@ -7,92 +7,165 @@
-        }
+  static void createTestCases()
+  {
+    const std::string yaml_file_path = ament_index_cpp::get_package_share_directory("moveit_middleware_benchmark") +
```
A good best practice is to put constants at the beginning of a file (after the includes) into an anonymous namespace

sjahr commented on test/test_scenario_perception_pipeline.cpp:
```diff
@@ -7,92 +7,165 @@
+const std::string PLANNING_GROUP = "panda_arm";
 
-class ScenarioPerceptionPipeline {
+class ScenarioPerceptionPipeline
```
I'd recommend putting this class into a namespace e.g. moveit::benchmarking

CihatAltiparmak replied:
I did. If it is not okay, please reopen this conservation.

sjahr reviewed (commented):
Thank you! I gave it a first high-level pass, still need to test it though. It is great that you have already developed so much and we can start refining it :+1: An important first step is to restructure the repository such that it follows the same patterns as MoveIt does and thus make it easier to review and iterate for all of us (see my comments)

CihatAltiparmak committed:
Fix CI for UPSTREAM_WORKSPACE

- Added upstream workspace envrionment to CI
- Changed ROS version in CI
- Fixed CMakeList.txt and package.xml based on the feedbacks of CI
- Fixed upstream repository urls
- Changed test cases for more convenient benchmarking

CihatAltiparmak committed:
Rearranged directory tree structure and some modifications

- Added middleware implementations to exec_depend ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/1#discussion_r1647678198 )
- Refactored package directory tree structure ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/1#discussion_r1647697922 )
- Added namespace ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/1#discussion_r1647703131 )

CihatAltiparmak commented:
> An important first step is to restructure the repository such that it follows the same patterns as MoveIt does and thus make it easier to review and iterate for all of us (see my comments)

Is there any documentation of how to create a new moveit repository and what its procedure is? Or is there any boilerplate to create moveit-like repository? If the answers are yes, maybe i can help you to spend the less effort by applying this steps.

CihatAltiparmak committed:
Added documentation and fixed package.xml

- Added documentation for each methods and variables except global variables ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/1#discussion_r1647698884 )
- Put constants into anoymous namespace ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/1#discussion_r1647701794 )
- Modified condition checking in senfTargetPose ( https://github.com/CihatAltiparmak/moveit_middleware_benchmark/pull/1#discussion_r1647681828 )
- Fixed middleware packages in package.xml
- Renamed the test case config

CihatAltiparmak committed:
Removed ScenarioPerceptionPipelineTestCaseGenerator class

- Removed ScenarioPerceptionPipelineTestCaseGenerator class and added its methods to ScenarioPerceptionPipelineFixture
- Renamed ScenarioPerceptionPipelineFixture/selectTestCase method as ScenarioPerceptionPipelineFixture/selectTestCase
- Added documentation for ScenarioPerceptionPipelineFixture/setUp and ScenarioPerceptionPipelineFixture/tearDown methods
